DOCS-7855-7856-4.3-Changes-for-SDK (#63)
mulesoft/docs-mule-sdk: modules/ROOT/pages/binary-streaming.adoc

</file:read>

> ----

Reviewer: Remainder of PR edit = copyedits only, of changed text only

== Stream from Existing Files

Available since version 1.3.

> When the `<repeatable-file-store-stream>` strategy is used, a file is used to temporarily hold information that doesn’t fit into the in-memory buffer. There are particular cases in which this is inefficient and leads to suboptimal IO and disk space requirements.

Reviewer: Intentional passive voice?

Reviewer: To avoid two "is used"s in close prox:
```suggestion
When the `<repeatable-file-store-stream>` strategy is used, a file temporarily holds information that doesn’t fit into the in-memory buffer. There are particular cases in which this is inefficient and leads to suboptimal IO and disk space requirements.
```
Does "IO" here refer to IO performance or IO requirements?

Reviewer 2: IO is the common term for Input/Output

Author: ```suggestion
When the `<repeatable-file-store-stream>` strategy is used, a file temporarily holds information that doesn’t fit into the in-memory buffer. There are particular cases in which using this strategy is inefficient and leads to suboptimal I/O and disk space requirements.
```

> Most common cases of this include:

Reviewer: Clarify "this": when "using this file" this is common? When "using this file? proves suboptimal?

Reviewer: (Oops...okay, copyediting only...:)

Author: ```suggestion
Most common scenarios for streaming from existing files include:
```

Author: NTS: (make sure we want this change)

Author: ```suggestion
The most common cases of this include:
```

Scenario 1::

> You are building an Operation or Source that returns the content of a file. The component that you developed opens a file and returns an InputStream that contains the file content. If the file is bigger than the memory buffer, the <repeatable-file-store-stream> starts building a temporary file with the stream content. That file becomes an exact duplicate of the original file that was already opened

Reviewer: "you are building" vs. "you developed" vs. "becomes": in which tense is this scenario?

Reviewer: End with a period.

Author: ```suggestion
You are building an Operation or Source that returns the content of a file. The component that you are developing opens a file and returns an InputStream that contains the file content. If the file is bigger than the memory buffer, the <repeatable-file-store-stream> starts building a temporary file with the stream content. That file becomes an exact duplicate of the original file that was already opened.
```

Scenario 2::

> There is a variation of the above in which an Operation or Source that obtains an InputStream and needs to pre-process it *before* it is returned. Because the original stream is not yet pushed to the flow, it hasn’t been made repeatable. As a result, the source needs to implement its own buffering by storing the processed contents into a temporary file as it processed the original stream. When pre-processing finishes, the source then pushes the contents of the file, going back to scenario 1.

Reviewer: ```suggestion
In a variation of Scenario 1, an Operation or Source obtains an InputStream and needs to process it _before_ it is returned. Because the original stream is not yet pushed to the flow, it hasn’t been made repeatable. As a result, the source needs to implement its own buffering by storing the processed contents into a temporary file at the same time that it processes the original stream. When the original stream processing finishes, the source then pushes the contents of the file, going back to Scenario 1.
```
Do we need "preprocess" when we follow it with "before"? Doesn't "before"= preprocessing?
The source stores the (pre)processed content while processing it, or while processing the whole stream?
When preprocessing finishes, or when processing of the original stream finishes?

Reviewer 2: This suggestion breaks the meaning of the paragraph. There's a typo in the original content however: where it says `contents into a temporary file as it processed the original stream` it should say `contents into a temporary file as it processes the original stream`

Author: ```suggestion
In a variation of Scenario 1, an Operation or Source obtains an InputStream and needs to pre-process it _before_ it is returned. Because the original stream is not yet pushed to the flow, it hasn’t been made repeatable. As a result, the source needs to implement its own buffering by storing the processed contents into a temporary file as it processes the original stream. When pre-processing finishes, the source then pushes the contents of the file, going back to Scenario 1.
```

Author: NTS verify

> The above scenarios articulate a need to get a repeatable stream from the content of an existing file, without the repeatable streaming framework creating a new file of its own, since the contents already originate from a file.

Reviewer: ```suggestion
These two scenarios articulate a need to get a repeatable stream from the content of an existing file, without the repeatable streaming framework creating a new file of its own, since the contents already originate from a file.
```

Reviewer: "...creating a new file of its own and thereby unnecessarily duplicating the file that already exists."?

Author: ```suggestion
These two scenarios articulate a need to get a repeatable stream from the content of an existing file, without the repeatable streaming framework creating a new file of its own, since the contents already originate from a file.
```

Author: NTS, this might be redundant. check.

> For this use case, you can use the `ImmutableFileRepeatableInputStream` class. This class is similar to the standard `java.io.FileInputStream`, but it instructs the Mule Runtime not to generate a new buffering file but to reuse the one that already exists.

Reviewer: ```suggestion
In these scenarios, use the `ImmutableFileRepeatableInputStream` class, which is similar to the standard `java.io.FileInputStream` but prevents Mule runtime engine (Mule) from generating a new buffering file and instructs Mule to reuse the existing file.
```

> For example, it can be used in an operation:

Reviewer: ```suggestion
For example, this class can be used in an operation:
```
Intentionally passive voice?

[source, java, linenums]
----
@MediaType(value = ANY, strict = false)
public InputStream read(String path) {
	File file = new File(path);
	if (!file.exists()) {
		// throw ModuleException
	}

	return new ImmutableFileRepeatableInputStream(file);
}
----

> An important aspect of dealing with files of this nature is understanding when to delete them. Depending on the use case, you might need to do that on your own code. In simpler cases, you can delete it when the stream is closed. If you want to delete the file when the stream is closed, you can use the following convenience constructor that allows you to instruct the Runtime to automatically delete the file when the stream is closed:

Reviewer: ```suggestion
An important aspect of working with files of this nature is understanding when to delete them. Depending on the use case, you might need to delete them through your own code. In simpler cases, you can delete files when the stream is closed. If you want to delete the file when the stream is closed, you can use the following convenience constructor, which enables you to instruct Mule to automatically delete the file when the stream is closed:
```

> `new ImmutableFileRepeatableInputStream(file, true);``

Author: ```suggestion
`new ImmutableFileRepeatableInputStream(file, true);`
```

[IMPORTANT]

> The file must be immutable in order for you to use this feature. That means that once the ImmutableFileRepeatableInputStream instance is created, the contents of that file cannot change, at least until the ImmutableFileRepeatableInputStream is closed. Changing the file’s content while the flow is reading it can lead to inconsistencies, dirty reads, and other types of race conditions.

Reviewer: ```suggestion
The file must be immutable for you to use this feature. That means that once the `ImmutableFileRepeatableInputStream` instance is created, the contents of that file cannot change, at least until `ImmutableFileRepeatableInputStream` is closed. Changing the file’s content while the flow is reading it can lead to inconsistencies, dirty reads, and other types of race conditions.
```

== See Also

* xref:streaming.adoc[Streaming]

> * xref:mule-runtime::4.1::streaming-about.adoc#repeatable_streams[Repeatable Streams]

Author: Note to self: run link checker, especially for this puppy.
